File: backend/app/schemas/notify_preferences.py
"""
通知偏好 Schema
NOTIFY-UX-1 实现
"""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field

from app.models.enums.notification_type import NotificationType
from app.models.enums.reading_media_type import ReadingMediaType


# ============== UserNotifyPreference Schema ==============

class UserNotifyPreferenceBase(BaseModel):
    """通知偏好基础 Schema"""
    notification_type: NotificationType
    media_type: Optional[ReadingMediaType] = None
    target_id: Optional[int] = None
    enable_web: bool = True
    enable_telegram: bool = True
    enable_webhook: bool = True
    enable_bark: bool = True
    muted: bool = False
    digest_only: bool = False


class UserNotifyPreferenceCreate(UserNotifyPreferenceBase):
    """创建通知偏好"""
    pass


class UserNotifyPreferenceUpdate(BaseModel):
    """更新通知偏好"""
    notification_type: NotificationType
    media_type: Optional[ReadingMediaType] = None
    target_id: Optional[int] = None
    enable_web: Optional[bool] = None
    enable_telegram: Optional[bool] = None
    enable_webhook: Optional[bool] = None
    enable_bark: Optional[bool] = None
    muted: Optional[bool] = None
    digest_only: Optional[bool] = None


class UserNotifyPreferenceRead(UserNotifyPreferenceBase):
    """读取通知偏好"""
    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ============== UserNotifySnooze Schema ==============

class UserNotifySnoozeBase(BaseModel):
    """通知静音基础 Schema"""
    muted: bool = False
    snooze_until: Optional[datetime] = None
    allow_critical_only: bool = False


class UserNotifySnoozeUpdate(BaseModel):
    """更新静音状态"""
    muted: Optional[bool] = None
    snooze_until: Optional[datetime] = None
    allow_critical_only: Optional[bool] = None


class UserNotifySnoozeRead(UserNotifySnoozeBase):
    """读取静音状态"""
    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ============== 聚合 DTO ==============

class NotificationTypeInfo(BaseModel):
    """通知类型信息"""
    type: NotificationType
    name: str  # 显示名称
    description: str  # 描述
    group: str  # 分组（manga/novel/music/system/task）
    is_critical: bool = False  # 是否为重要通知


class UserNotifyPreferenceMatrix(BaseModel):
    """用户通知偏好矩阵（聚合）"""
    preferences: list[UserNotifyPreferenceRead]
    snooze: Optional[UserNotifySnoozeRead] = None
    available_notification_types: list[NotificationTypeInfo]


# ============== 投递决策 ==============

class NotificationDeliveryDecision(BaseModel):
    """通知投递决策"""
    allowed_web: bool = True
    allowed_telegram: bool = True
    allowed_webhook: bool = True
    allowed_bark: bool = True
    store_in_user_notification: bool = True  # 是否写入 UserNotification 表
    reason: Optional[str] = None  # 决策原因（调试用）


# ============== 快捷操作 ==============

class MuteNotificationTypeRequest(BaseModel):
    """静音某类通知请求"""
    notification_type: NotificationType
    media_type: Optional[ReadingMediaType] = None
    target_id: Optional[int] = None


class SnoozeRequest(BaseModel):
    """Snooze 请求"""
    duration_minutes: Optional[int] = Field(None, ge=5, le=1440, description="静音时长（分钟），5-1440")
    until: Optional[datetime] = Field(None, description="静音到指定时间")
    allow_critical_only: bool = False

File: frontend/src/types/notifyPreferences.ts
/**
 * 通知偏好类型定义
 * NOTIFY-UX-1 实现
 */

// 通知类型枚举
export type NotificationType =
  | 'MANGA_NEW_CHAPTER'
  | 'MANGA_UPDATED'
  | 'MANGA_SYNC_FAILED'
  | 'NOVEL_NEW_CHAPTER'
  | 'AUDIOBOOK_NEW_TRACK'
  | 'TTS_JOB_COMPLETED'
  | 'TTS_JOB_FAILED'
  | 'AUDIOBOOK_READY'
  | 'MUSIC_CHART_UPDATED'
  | 'MUSIC_NEW_TRACKS_QUEUED'
  | 'MUSIC_NEW_TRACKS_DOWNLOADING'
  | 'MUSIC_NEW_TRACKS_READY'
  | 'SYSTEM_MESSAGE'

// 媒体类型
export type ReadingMediaType = 'NOVEL' | 'AUDIOBOOK' | 'MANGA'

// 通知类型信息
export interface NotificationTypeInfo {
  type: NotificationType
  name: string
  description: string
  group: string // manga / novel / music / system / task
  is_critical: boolean
}

// 用户通知偏好
export interface UserNotifyPreference {
  id: number
  user_id: number
  notification_type: NotificationType
  media_type?: ReadingMediaType | null
  target_id?: number | null
  enable_web: boolean
  enable_telegram: boolean
  enable_webhook: boolean
  enable_bark: boolean
  muted: boolean
  digest_only: boolean
  created_at: string
  updated_at: string
}

// 用户静音状态
export interface UserNotifySnooze {
  id: number
  user_id: number
  muted: boolean
  snooze_until?: string | null
  allow_critical_only: boolean
  created_at: string
  updated_at: string
}

// 偏好矩阵（聚合）
export interface UserNotifyPreferenceMatrix {
  preferences: UserNotifyPreference[]
  snooze?: UserNotifySnooze | null
  available_notification_types: NotificationTypeInfo[]
}

// 创建/更新偏好请求
export interface UserNotifyPreferenceUpsert {
  notification_type: NotificationType
  media_type?: ReadingMediaType | null
  target_id?: number | null
  enable_web?: boolean
  enable_telegram?: boolean
  enable_webhook?: boolean
  enable_bark?: boolean
  muted?: boolean
  digest_only?: boolean
}

// 更新静音请求
export interface UserNotifySnoozeUpdate {
  muted?: boolean
  snooze_until?: string | null
  allow_critical_only?: boolean
}

// 快速 Snooze 请求
export interface SnoozeRequest {
  duration_minutes?: number // 5-1440
  until?: string
  allow_critical_only?: boolean
}

// 静音某类通知请求
export interface MuteNotificationTypeRequest {
  notification_type: NotificationType
  media_type?: ReadingMediaType | null
  target_id?: number | null
}

// 通知类型分组配置（前端展示用）
export const NOTIFICATION_TYPE_GROUPS: Record<string, {
  label: string
  icon: string
  color: string
}> = {
  manga: {
    label: '漫画',
    icon: 'mdi-book-open-page-variant',
    color: 'orange',
  },
  novel: {
    label: '小说/有声书',
    icon: 'mdi-book',
    color: 'blue',
  },
  music: {
    label: '音乐',
    icon: 'mdi-music',
    color: 'purple',
  },
  system: {
    label: '系统',
    icon: 'mdi-cog',
    color: 'grey',
  },
}
